starting container with HostConfig was deprecated since v1.10 and removed in v1.12

On a machine with docker v1.12 it throws above error error, but works fine on docker v1.10.3

running:
```bash
$ imagebuilder . surajkanginx
--> Image fedora was not found, pulling ...
--> Pulled 0/1 layers, 7% complete
--> Pulled 1/1 layers, 100% complete
--> Extracting
--> FROM fedora
unable to start build container: API error (400): {"message":"starting container with HostConfig was deprecated since v1.10 and removed in v1.12"}
```

dockerfile:
```bash
 $ cat Dockerfile 
FROM fedora
MAINTAINER http://fedoraproject.org/wiki/Cloud

RUN dnf -y update && dnf clean all
RUN dnf -y install nginx && dnf clean all
RUN echo "daemon off;" >> /etc/nginx/nginx.conf
RUN echo "nginx on Fedora" > /usr/share/nginx/html/index.html

EXPOSE 80

CMD [ "/usr/sbin/nginx" ]
```


docker version
```bash
$ docker version
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64
```

Unfortunately, the conformance tests are failing with 1.12. This requires further investigation, though at least it seems like a step in the right direction.

```
$ go test -v github.com/openshift/imagebuilder/dockerclient/  -tags conformance
=== RUN   TestConformanceInternal
--- PASS: TestConformanceInternal (28.46s)
=== RUN   TestConformanceExternal
--- PASS: TestConformanceExternal (485.71s)
	conformance_test.go:299: WARNING: opt/app-root/src/.gem/ruby/gems/json-1.8.3/ext/json/ext/parser/Makefile differs only in size by -2 bytes, probably a timestamp value change
	conformance_test.go:299: WARNING: opt/app-root/src/.gem/ruby/extensions/x86_64-linux/json-1.8.3/gem_make.out differs only in size by -1 bytes, probably a timestamp value change
	conformance_test.go:299: WARNING: opt/app-root/src/.gem/ruby/gems/json-1.8.3/ext/json/ext/generator/Makefile differs only in size by -2 bytes, probably a timestamp value change
	conformance_test.go:299: WARNING: etc/ssl/private/ssl-cert-snakeoil.key differs only in size by 4 bytes, probably a timestamp value change
=== RUN   TestTransientMount
--- FAIL: TestTransientMount (1.25s)
	conformance_test.go:167: unable to build image: running '/bin/sh -c ls /mountdir/subdir' failed with exit code 1
=== RUN   TestCalcCopyInfo
--- PASS: TestCalcCopyInfo (0.00s)
FAIL
exit status 1
FAIL	github.com/openshift/imagebuilder/dockerclient	515.426s
```

--- a/README.md
+++ b/README.md
@@ -40,10 +40,11 @@ To download and install the library and the binary, set up a Golang build enviro
 $ go get -u github.com/openshift/imagebuilder/cmd/imagebuilder
 ```
 
-The included command line takes two arguments, a path to a directory containing a dockerfile, and the name to tag the output image with.
+The included command line takes one argument, a path to a directory containing a Dockerfile. The `-t` option
+can be used to specify an image to tag as.
 
 ```
-$ imagebuilder path_to_directory output_image_name
+$ imagebuilder [-t TAG] DIRECTORY
 ```
 
 To mount a file into the image for build that will not be present in the final output image, run:
@@ -54,7 +55,18 @@ $ imagebuilder --mount ~/secrets/private.key:/etc/keys/private.key path/to/my/co
 
 Any processes in the Dockerfile will have access to `/etc/keys/private.key`, but that file will not be part of the committed image.
 
-Note that running `--mount` requires Docker 1.10 or newer, as it uses a Docker volume to hold the mounted files and the volume API was not available in earlier versions.
+Running `--mount` requires Docker 1.10 or newer, as it uses a Docker volume to hold the mounted files and the volume API was not
+available in earlier versions. In addition, there is currently a bug that blocks Docker 1.12 from using `--mount`.
+
+You can also customize which Dockerfile is run, or run multiple Dockerfiles in sequenc (the FROM is ignored on
+later files):
+
+```
+$ imagebuilder -f Dockerfile:Dockerfile.extra .
+```
+
+will build the current directory and combine the first Dockerfile with the second. The FROM in the second image
+is ignored.
 
 
 ## Code Example

--- a/builder.go
+++ b/builder.go
@@ -3,6 +3,7 @@ package imagebuilder
 import (
 	"bytes"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"log"
 	"os"
@@ -86,7 +87,7 @@ type Builder struct {
 	PendingRuns   []Run
 	PendingCopies []Copy
 
-	Executor Executor
+	//Executor Executor
 }
 
 func NewBuilder() *Builder {
@@ -98,6 +99,25 @@ func NewBuilder() *Builder {
 		Args:        make(map[string]string),
 		AllowedArgs: args,
 	}
+}
+
+func NewBuilderForReader(r io.Reader, args map[string]string) (*Builder, *parser.Node, error) {
+	b := NewBuilder()
+	b.Args = args
+	node, err := ParseDockerfile(r)
+	if err != nil {
+		return nil, nil, err
+	}
+	return b, node, err
+}
+
+func NewBuilderForFile(path string, args map[string]string) (*Builder, *parser.Node, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer f.Close()
+	return NewBuilderForReader(f, args)
 }
 
 // Step creates a new step from the current state.

--- a/cmd/imagebuilder/imagebuilder.go
+++ b/cmd/imagebuilder/imagebuilder.go
@@ -12,16 +12,21 @@ import (
 	docker "github.com/fsouza/go-dockerclient"
 	"github.com/golang/glog"
 
+	"github.com/openshift/imagebuilder"
 	"github.com/openshift/imagebuilder/dockerclient"
 )
 
 func main() {
 	log.SetFlags(0)
 	options := dockerclient.NewClientExecutor(nil)
+	var tag string
 	var dockerfilePath string
 	var mountSpecs stringSliceFlag
 
-	flag.StringVar(&dockerfilePath, "dockerfile", dockerfilePath, "An optional path to a Dockerfile to use.")
+	flag.StringVar(&tag, "t", tag, "The name to assign this image, if any.")
+	flag.StringVar(&tag, "tag", tag, "The name to assign this image, if any.")
+	flag.StringVar(&dockerfilePath, "f", dockerfilePath, "An optional path to a Dockerfile to use. You may pass multiple docker files using the operating system delimiter.")
+	flag.StringVar(&dockerfilePath, "file", dockerfilePath, "An optional path to a Dockerfile to use. You may pass multiple docker files using the operating system delimiter.")
 	flag.Var(&mountSpecs, "mount", "An optional list of files and directories to mount during the build. Use SRC:DST syntax for each path.")
 	flag.BoolVar(&options.AllowPull, "allow-pull", true, "Pull the images that are not present.")
 	flag.BoolVar(&options.IgnoreUnrecognizedInstructions, "ignore-unrecognized-instructions", true, "If an unrecognized Docker instruction is encountered, warn but do not fail the build.")
@@ -29,12 +34,12 @@ func main() {
 	flag.Parse()
 
 	args := flag.Args()
-	if len(args) != 2 {
-		log.Fatalf("You must provide two arguments: DIRECTORY and IMAGE_NAME")
+	if len(args) != 1 {
+		log.Fatalf("You must provide one argument, the name of a directory to build")
 	}
 
 	options.Directory = args[0]
-	options.Tag = args[1]
+	options.Tag = tag
 	if len(dockerfilePath) == 0 {
 		dockerfilePath = filepath.Join(options.Directory, "Dockerfile")
 	}
@@ -64,30 +69,56 @@ func main() {
 	// Accept ARGS on the command line
 	arguments := make(map[string]string)
 
-	if err := build(dockerfilePath, options, arguments); err != nil {
+	dockerfiles := filepath.SplitList(dockerfilePath)
+	if len(dockerfiles) == 0 {
+		dockerfiles = []string{filepath.Join(options.Directory, "Dockerfile")}
+	}
+
+	if err := build(dockerfiles[0], dockerfiles[1:], arguments, options); err != nil {
 		log.Fatal(err.Error())
 	}
 }
 
-func build(dockerfilePath string, e *dockerclient.ClientExecutor, arguments map[string]string) error {
+func build(dockerfile string, additionalDockerfiles []string, arguments map[string]string, e *dockerclient.ClientExecutor) error {
+	if err := e.DefaultExcludes(); err != nil {
+		return fmt.Errorf("error: Could not parse default .dockerignore: %v", err)
+	}
+
 	client, err := docker.NewClientFromEnv()
 	if err != nil {
-		log.Fatalf("No connection to Docker available: %v", err)
+		return fmt.Errorf("error: No connection to Docker available: %v", err)
 	}
 	e.Client = client
 
-	f, err := os.Open(dockerfilePath)
+	// TODO: handle signals
+	defer func() {
+		for _, err := range e.Release() {
+			fmt.Fprintf(e.ErrOut, "error: Unable to clean up build: %v\n", err)
+		}
+	}()
+
+	b, node, err := imagebuilder.NewBuilderForFile(dockerfile, arguments)
 	if err != nil {
 		return err
 	}
-	defer f.Close()
-
-	// TODO: handle signals
-	if err := e.Cleanup(e.Container); err != nil {
-		fmt.Fprintf(e.ErrOut, "error: Unable to clean up build: %v\n", err)
+	if err := e.Prepare(b, node); err != nil {
+		return err
+	}
+	if err := e.Execute(b, node); err != nil {
+		return err
 	}
 
-	return e.Build(f, arguments)
+	for _, s := range additionalDockerfiles {
+		_, node, err := imagebuilder.NewBuilderForFile(s, arguments)
+		if err != nil {
+			return err
+		}
+		if err := e.Execute(b, node); err != nil {
+			return err
+		}
+	}
+
+	return e.Commit(b)
 }
 
 type stringSliceFlag []string

--- a/cmd/imagebuilder/imagebuilder.go
+++ b/cmd/imagebuilder/imagebuilder.go
@@ -17,7 +17,7 @@ import (
 
 func main() {
 	log.SetFlags(0)
-	options := &dockerclient.ClientExecutor{}
+	options := dockerclient.NewClientExecutor(nil)
 	var dockerfilePath string
 	var mountSpecs stringSliceFlag
 
@@ -83,7 +83,7 @@ func build(dockerfilePath string, e *dockerclient.ClientExecutor, arguments map[
 	defer f.Close()
 
 	// TODO: handle signals
-	if err := e.Cleanup(); err != nil {
+	if err := e.Cleanup(e.Container); err != nil {
 		fmt.Fprintf(e.ErrOut, "error: Unable to clean up build: %v\n", err)
 	}
 

--- a/dockerclient/client.go
+++ b/dockerclient/client.go
@@ -13,6 +13,7 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/docker/docker/builder/dockerfile/parser"
 	"github.com/docker/docker/pkg/archive"
 	"github.com/docker/docker/pkg/fileutils"
 	dockertypes "github.com/docker/engine-api/types"
@@ -78,6 +79,10 @@ type ClientExecutor struct {
 	HostConfig *docker.HostConfig
 	// LogFn is an optional command to log information to the end user
 	LogFn func(format string, args ...interface{})
+
+	// Deferred is a list of operations that must be cleaned up at
+	// the end of execution. Use Release() to handle these.
+	Deferred []func() error
 }
 
 // NewClientExecutor creates a client executor.
@@ -90,31 +95,33 @@ func NewClientExecutor(client *docker.Client) *ClientExecutor {
 	}
 }
 
+func (e *ClientExecutor) DefaultExcludes() error {
+	excludes, err := imagebuilder.ParseDockerignore(e.Directory)
+	if err != nil {
+		return err
+	}
+	e.Excludes = append(excludes, ".dockerignore")
+	return nil
+}
+
 // Build is a helper method to perform a Docker build against the
 // provided Docker client. It will load the image if not specified,
 // create a container if one does not already exist, and start a
 // container if the Dockerfile contains RUN commands. It will cleanup
 // any containers it creates directly, and set the e.Image.ID field
 // to the generated image.
-func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
-	b := imagebuilder.NewBuilder()
-	b.Args = args
-
-	if e.Excludes == nil {
-		excludes, err := imagebuilder.ParseDockerignore(e.Directory)
-		if err != nil {
-			return err
-		}
-		e.Excludes = append(excludes, ".dockerignore")
-	}
-
-	// TODO: check the Docker daemon version (1.20 is required for Upload)
-
-	node, err := imagebuilder.ParseDockerfile(r)
-	if err != nil {
+func (e *ClientExecutor) Build(b *imagebuilder.Builder, node *parser.Node) error {
+	defer e.Release()
+	if err := e.Prepare(b, node); err != nil {
 		return err
 	}
+	if err := e.Execute(b, node); err != nil {
+		return err
+	}
+	return e.Commit(b)
+}
 
+func (e *ClientExecutor) Prepare(b *imagebuilder.Builder, node *parser.Node) error {
 	// identify the base image
 	from, err := b.From(node)
 	if err != nil {
@@ -130,7 +137,7 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 			if err != nil {
 				return fmt.Errorf("unable to create a scratch image for this build: %v", err)
 			}
-			defer e.CleanupImage(from)
+			e.Deferred = append(e.Deferred, func() error { return e.Client.RemoveImage(from) })
 		}
 		glog.V(4).Infof("Retrieving image %q", from)
 		e.Image, err = e.LoadImage(from)
@@ -176,10 +183,10 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 				if err != nil {
 					return fmt.Errorf("unable to create volume to mount secrets: %v", err)
 				}
-				defer e.cleanupVolume(volumeName)
+				e.Deferred = append(e.Deferred, func() error { return e.Client.RemoveVolume(volumeName) })
 				sharedMount = v.Mountpoint
 				opts.HostConfig = &docker.HostConfig{
-					Binds: []string{sharedMount + ":" + e.ContainerTransientMount},
+					Binds: []string{volumeName + ":" + e.ContainerTransientMount},
 				}
 			}
 
@@ -203,7 +210,7 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 			if len(sharedMount) == 0 {
 				return fmt.Errorf("no mount point available for temporary mounts")
 			}
-			binds, err := e.StartPartialContainer(opts, e.TransientMounts, sharedMount)
+			binds, err := e.PopulateTransientMounts(opts, e.TransientMounts, sharedMount)
 			if err != nil {
 				return err
 			}
@@ -215,9 +222,7 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 			return fmt.Errorf("unable to create build container: %v", err)
 		}
 		e.Container = container
-
-		// if we create the container, take responsibilty for cleaning up
-		defer e.Cleanup(container)
+		e.Deferred = append(e.Deferred, func() error { return e.removeContainer(container.ID) })
 	}
 
 	// TODO: lazy start
@@ -225,9 +230,15 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 		if err := e.Client.StartContainer(e.Container.ID, nil); err != nil {
 			return fmt.Errorf("unable to start build container: %v", err)
 		}
+		e.Container.State.Running = true
 		// TODO: is this racy? may have to loop wait in the actual run step
 	}
+	return nil
+}
 
+// Execute performs all of the provided steps against the initialized container. May be
+// invoked multiple times for a given container.
+func (e *ClientExecutor) Execute(b *imagebuilder.Builder, node *parser.Node) error {
 	for _, child := range node.Children {
 		step := b.Step()
 		if err := step.Resolve(child); err != nil {
@@ -242,15 +253,22 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 		}
 	}
 
-	if mustStart {
+	return nil
+}
+
+// Commit saves the completed build as an image with the provided tag. It will
+// stop the container, commit the image, and then remove the container.
+func (e *ClientExecutor) Commit(b *imagebuilder.Builder) error {
+	config := b.Config()
+
+	if e.Container.State.Running {
 		glog.V(4).Infof("Stopping container %s ...", e.Container.ID)
 		if err := e.Client.StopContainer(e.Container.ID, 0); err != nil {
 			return fmt.Errorf("unable to stop build container: %v", err)
 		}
-	}
-
-	config := b.Config()
-	if mustStart {
+		e.Container.State.Running = false
+		// Starting the container may perform escaping of args, so to be consistent
+		// we also set that here
 		config.ArgsEscaped = true
 	}
 
@@ -267,6 +285,8 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 			e.LogFn("Committing changes ...")
 		}
 	}
+
+	defer e.Release()
 
 	image, err := e.Client.CommitContainer(docker.CommitContainerOptions{
 		Author:     b.Author,
@@ -286,45 +306,54 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 	return nil
 }
 
-func (e *ClientExecutor) StartPartialContainer(opts docker.CreateContainerOptions, transientMounts []Mount, sharedMount string) ([]string, error) {
+func (e *ClientExecutor) PopulateTransientMounts(opts docker.CreateContainerOptions, transientMounts []Mount, sharedMount string) ([]string, error) {
 	container, err := e.Client.CreateContainer(opts)
 	if err != nil {
-		return nil, fmt.Errorf("unable to create build container: %v", err)
+		return nil, fmt.Errorf("unable to create transient container: %v", err)
 	}
-	defer e.Cleanup(container)
+	defer e.removeContainer(container.ID)
 
 	var copies []imagebuilder.Copy
 	for i, mount := range transientMounts {
 		source := mount.SourcePath
 		copies = append(copies, imagebuilder.Copy{
 			Src:  source,
-			Dest: []string{path.Join(e.ContainerTransientMount, strconv.Itoa(i))},
+			Dest: []string{filepath.Join(e.ContainerTransientMount, strconv.Itoa(i))},
 		})
 	}
 	if err := e.CopyContainer(container, copies...); err != nil {
-		return nil, fmt.Errorf("unable to copy build context into container: %v", err)
+		return nil, fmt.Errorf("unable to copy transient context into container: %v", err)
 	}
 
 	// mount individual items temporarily
 	var binds []string
 	for i, mount := range e.TransientMounts {
-		binds = append(binds, fmt.Sprintf("%s:%s:%s", path.Join(sharedMount, strconv.Itoa(i)), mount.DestinationPath, "ro"))
+		binds = append(binds, fmt.Sprintf("%s:%s:%s", filepath.Join(sharedMount, strconv.Itoa(i)), mount.DestinationPath, "ro"))
 	}
 	return binds, nil
 }
 
-// Cleanup will remove the container that created the build.
-func (e *ClientExecutor) Cleanup(container *docker.Container) error {
-	if container == nil {
-		return nil
+// Release deletes any items started by this executor.
+func (e *ClientExecutor) Release() []error {
+	var errs []error
+	for _, fn := range e.Deferred {
+		if err := fn(); err != nil {
+			errs = append(errs, err)
+		}
 	}
+	e.Deferred = nil
+	return errs
+}
+
+// removeContainer removes the provided container ID
+func (e *ClientExecutor) removeContainer(id string) error {
 	err := e.Client.RemoveContainer(docker.RemoveContainerOptions{
-		ID:            container.ID,
+		ID:            id,
 		RemoveVolumes: true,
 		Force:         true,
 	})
 	if _, ok := err.(*docker.NoSuchContainer); err != nil && !ok {
-		return fmt.Errorf("unable to cleanup build container: %v", err)
+		return fmt.Errorf("unable to cleanup container: %v", err)
 	}
 	return nil
 }
@@ -367,16 +396,6 @@ func randSeq(source string, n int) (string, error) {
 		random[i] = source[random[i]%byte(len(source))]
 	}
 	return string(random), nil
-}
-
-// cleanupVolume attempts to remove the provided volume
-func (e *ClientExecutor) cleanupVolume(name string) error {
-	return e.Client.RemoveVolume(name)
-}
-
-// CleanupImage attempts to remove the provided image.
-func (e *ClientExecutor) CleanupImage(name string) error {
-	return e.Client.RemoveImage(name)
 }
 
 // LoadImage checks the client for an image matching from. If not found,
@@ -517,6 +536,7 @@ func (e *ClientExecutor) CopyContainer(container *docker.Container, copies ...im
 			if err != nil {
 				return err
 			}
+
 			glog.V(5).Infof("Uploading to %s at %s", container.ID, dst)
 			err = e.Client.UploadToContainer(container.ID, docker.UploadToContainerOptions{
 				InputStream: r,

--- a/dockerclient/client.go
+++ b/dockerclient/client.go
@@ -54,6 +54,10 @@ type ClientExecutor struct {
 	// omitted from the final image.
 	TransientMounts []Mount
 
+	// The path within the container to perform the transient mount.
+	ContainerTransientMount string
+
+	// The streams used for canonical output.
 	Out, ErrOut io.Writer
 
 	// Container is optional and can be set to a container to use as
@@ -81,6 +85,8 @@ func NewClientExecutor(client *docker.Client) *ClientExecutor {
 	return &ClientExecutor{
 		Client: client,
 		LogFn:  func(string, ...interface{}) {},
+
+		ContainerTransientMount: "/.imagebuilder-transient-mount",
 	}
 }
 
@@ -152,7 +158,13 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 			Config: &docker.Config{
 				Image: from,
 			},
+			HostConfig: &docker.HostConfig{},
 		}
+		if e.HostConfig != nil {
+			opts.HostConfig = e.HostConfig
+		}
+		originalBinds := opts.HostConfig.Binds
+
 		if mustStart {
 			// Transient mounts only make sense on images that will be running processes
 			if len(e.TransientMounts) > 0 {
@@ -167,7 +179,7 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 				defer e.cleanupVolume(volumeName)
 				sharedMount = v.Mountpoint
 				opts.HostConfig = &docker.HostConfig{
-					Binds: []string{sharedMount + ":/tmp/__temporarymount"},
+					Binds: []string{sharedMount + ":" + e.ContainerTransientMount},
 				}
 			}
 
@@ -181,9 +193,23 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 				opts.Config.Entrypoint = []string{"/bin/sh", "-c"}
 			}
 		}
+
 		if len(opts.Config.Cmd) == 0 {
 			opts.Config.Entrypoint = []string{"/bin/sh", "-c", "# NOP"}
 		}
+
+		// copy any source content into the temporary mount path
+		if mustStart && len(e.TransientMounts) > 0 {
+			if len(sharedMount) == 0 {
+				return fmt.Errorf("no mount point available for temporary mounts")
+			}
+			binds, err := e.StartPartialContainer(opts, e.TransientMounts, sharedMount)
+			if err != nil {
+				return err
+			}
+			opts.HostConfig.Binds = append(originalBinds, binds...)
+		}
+
 		container, err := e.Client.CreateContainer(opts)
 		if err != nil {
 			return fmt.Errorf("unable to create build container: %v", err)
@@ -191,43 +217,12 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 		e.Container = container
 
 		// if we create the container, take responsibilty for cleaning up
-		defer e.Cleanup()
-	}
-
-	// copy any source content into the temporary mount path
-	if mustStart && len(e.TransientMounts) > 0 {
-		var copies []imagebuilder.Copy
-		for i, mount := range e.TransientMounts {
-			source := mount.SourcePath
-			copies = append(copies, imagebuilder.Copy{
-				Src:  source,
-				Dest: []string{path.Join("/tmp/__temporarymount", strconv.Itoa(i))},
-			})
-		}
-		if err := e.Copy(copies...); err != nil {
-			return fmt.Errorf("unable to copy build context into container: %v", err)
-		}
+		defer e.Cleanup(container)
 	}
 
 	// TODO: lazy start
 	if mustStart && !e.Container.State.Running {
-		var hostConfig docker.HostConfig
-		if e.HostConfig != nil {
-			hostConfig = *e.HostConfig
-		}
-
-		// mount individual items temporarily
-		for i, mount := range e.TransientMounts {
-			if len(sharedMount) == 0 {
-				return fmt.Errorf("no mount point available for temporary mounts")
-			}
-			hostConfig.Binds = append(
-				hostConfig.Binds,
-				fmt.Sprintf("%s:%s:%s", path.Join(sharedMount, strconv.Itoa(i)), mount.DestinationPath, "ro"),
-			)
-		}
-
-		if err := e.Client.StartContainer(e.Container.ID, &hostConfig); err != nil {
+		if err := e.Client.StartContainer(e.Container.ID, nil); err != nil {
 			return fmt.Errorf("unable to start build container: %v", err)
 		}
 		// TODO: is this racy? may have to loop wait in the actual run step
@@ -291,20 +286,46 @@ func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
 	return nil
 }
 
+func (e *ClientExecutor) StartPartialContainer(opts docker.CreateContainerOptions, transientMounts []Mount, sharedMount string) ([]string, error) {
+	container, err := e.Client.CreateContainer(opts)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create build container: %v", err)
+	}
+	defer e.Cleanup(container)
+
+	var copies []imagebuilder.Copy
+	for i, mount := range transientMounts {
+		source := mount.SourcePath
+		copies = append(copies, imagebuilder.Copy{
+			Src:  source,
+			Dest: []string{path.Join(e.ContainerTransientMount, strconv.Itoa(i))},
+		})
+	}
+	if err := e.CopyContainer(container, copies...); err != nil {
+		return nil, fmt.Errorf("unable to copy build context into container: %v", err)
+	}
+
+	// mount individual items temporarily
+	var binds []string
+	for i, mount := range e.TransientMounts {
+		binds = append(binds, fmt.Sprintf("%s:%s:%s", path.Join(sharedMount, strconv.Itoa(i)), mount.DestinationPath, "ro"))
+	}
+	return binds, nil
+}
+
 // Cleanup will remove the container that created the build.
-func (e *ClientExecutor) Cleanup() error {
-	if e.Container == nil {
+func (e *ClientExecutor) Cleanup(container *docker.Container) error {
+	if container == nil {
 		return nil
 	}
 	err := e.Client.RemoveContainer(docker.RemoveContainerOptions{
-		ID:            e.Container.ID,
+		ID:            container.ID,
 		RemoveVolumes: true,
 		Force:         true,
 	})
 	if _, ok := err.(*docker.NoSuchContainer); err != nil && !ok {
 		return fmt.Errorf("unable to cleanup build container: %v", err)
 	}
-	e.Container = nil
 	return nil
 }
 
@@ -481,8 +502,13 @@ func (e *ClientExecutor) Run(run imagebuilder.Run, config docker.Config) error {
 	return nil
 }
 
+// Copy implements the executor copy function.
 func (e *ClientExecutor) Copy(copies ...imagebuilder.Copy) error {
-	container := e.Container
+	return e.CopyContainer(e.Container, copies...)
+}
+
+// CopyContainer copies the provided content into a destination container.
+func (e *ClientExecutor) CopyContainer(container *docker.Container, copies ...imagebuilder.Copy) error {
 	for _, c := range copies {
 		// TODO: reuse source
 		for _, dst := range c.Dest {

--- a/dockerclient/conformance_test.go
+++ b/dockerclient/conformance_test.go
@@ -9,6 +9,7 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"math/rand"
 	"os"
 	"os/exec"
 	"path/filepath"
@@ -140,6 +141,81 @@ func TestConformanceExternal(t *testing.T) {
 	for i, test := range testCases {
 		conformanceTester(t, c, test, i, *compareLayers)
 	}
+}
+
+func TestTransientMount(t *testing.T) {
+	c, err := docker.NewClientFromEnv()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	e := NewClientExecutor(c)
+	e.AllowPull = true
+	e.Directory = "testdata"
+	e.TransientMounts = []Mount{
+		{SourcePath: "dir", DestinationPath: "/mountdir"},
+		{SourcePath: "Dockerfile.env", DestinationPath: "/mountfile"},
+	}
+	e.Tag = fmt.Sprintf("conformance%d", rand.Int63())
+
+	defer c.RemoveImage(e.Tag)
+
+	out := &bytes.Buffer{}
+	e.Out = out
+	if err := e.Build(bytes.NewBufferString("FROM busybox\nRUN ls /mountdir/subdir\nRUN cat /mountfile\n"), nil); err != nil {
+		t.Fatalf("unable to build image: %v", err)
+	}
+	if !strings.Contains(out.String(), "ENV name=value\n") {
+		t.Errorf("did not find expected output:\n%s", out.String())
+	}
+	if !strings.Contains(out.String(), "file2\n") {
+		t.Errorf("did not find expected output:\n%s", out.String())
+	}
+
+	result, err := testContainerOutput(c, e.Tag, []string{"/bin/sh", "-c", "ls -al /mountdir"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if strings.Contains(result, "subdir") {
+		t.Errorf("did not find expected output:\n%s", result)
+	}
+	result, err = testContainerOutput(c, e.Tag, []string{"/bin/sh", "-c", "cat /mountfile"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if strings.Contains(result, "ENV name=value\n") {
+		t.Errorf("did not find expected output:\n%s", result)
+	}
+}
+
+func testContainerOutput(c *docker.Client, tag string, command []string) (string, error) {
+	container, err := c.CreateContainer(docker.CreateContainerOptions{
+		Name: tag + "-test",
+		Config: &docker.Config{
+			Image:      tag,
+			Entrypoint: command,
+			Cmd:        nil,
+		},
+	})
+	if err != nil {
+		return "", err
+	}
+	defer c.RemoveContainer(docker.RemoveContainerOptions{ID: container.ID})
+	if err := c.StartContainer(container.ID, nil); err != nil {
+		return "", err
+	}
+	code, err := c.WaitContainer(container.ID)
+	if err != nil {
+		return "", err
+	}
+	if code != 0 {
+		return "", fmt.Errorf("unrecognized exit code: %d", code)
+	}
+	out := &bytes.Buffer{}
+	if err := c.Logs(docker.LogsOptions{Container: container.ID, Stdout: true, OutputStream: out}); err != nil {
+		return "", fmt.Errorf("unable to get logs: %v", err)
+	}
+	return out.String(), nil
 }
 
 func conformanceTester(t *testing.T, c *docker.Client, test conformanceTest, i int, deep bool) {

--- a/dockerclient/conformance_test.go
+++ b/dockerclient/conformance_test.go
@@ -162,7 +162,11 @@ func TestTransientMount(t *testing.T) {
 
 	out := &bytes.Buffer{}
 	e.Out = out
-	if err := e.Build(bytes.NewBufferString("FROM busybox\nRUN ls /mountdir/subdir\nRUN cat /mountfile\n"), nil); err != nil {
+	b, node, err := imagebuilder.NewBuilderForReader(bytes.NewBufferString("FROM busybox\nRUN ls /mountdir/subdir\nRUN cat /mountfile\n"), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := e.Build(b, node); err != nil {
 		t.Fatalf("unable to build image: %v", err)
 	}
 	if !strings.Contains(out.String(), "ENV name=value\n") {
@@ -342,7 +346,11 @@ func conformanceTester(t *testing.T, c *docker.Client, test conformanceTest, i i
 			e.Out, e.ErrOut = out, out
 			e.Directory = contextDir
 			e.Tag = nameDirect
-			if err := e.Build(bytes.NewBufferString(testFile), nil); err != nil {
+			b, node, err := imagebuilder.NewBuilderForReader(bytes.NewBufferString(testFile), nil)
+			if err != nil {
+				t.Fatalf("%d: %v", i, err)
+			}
+			if err := e.Build(b, node); err != nil {
 				t.Errorf("%d: failed to build step %d in dockerfile %q: %s\n%s", i, j, dockerfilePath, steps[j].Original, out)
 				break
 			}
@@ -401,7 +409,11 @@ func conformanceTester(t *testing.T, c *docker.Client, test conformanceTest, i i
 		e.Out, e.ErrOut = out, out
 		e.Directory = contextDir
 		e.Tag = nameDirect
-		if err := e.Build(bytes.NewBuffer(data), nil); err != nil {
+		b, node, err := imagebuilder.NewBuilderForReader(bytes.NewBuffer(data), nil)
+		if err != nil {
+			t.Fatalf("%d: %v", i, err)
+		}
+		if err := e.Build(b, node); err != nil {
 			t.Errorf("%d: failed to build complete image in %q: %v\n%s", i, input, err, out)
 		} else {
 			if !equivalentImages(

